bindings/matrix-sdk-ffi/src/timeline/content.rs

```diff
@@ -74,45 +69,74 @@ impl TimelineItemContent {
                         )
                     })
                     .unzip();
-                TimelineItemContentKind::ProfileChange {
+                TimelineItemContent::ProfileChange {
                     display_name: display_name.flatten(),
                     prev_display_name: prev_display_name.flatten(),
                     avatar_url: avatar_url.flatten(),
                     prev_avatar_url: prev_avatar_url.flatten(),
                 }
             }
 
-            Content::OtherState(state) => TimelineItemContentKind::State {
+            Content::OtherState(state) => TimelineItemContent::State {
                 state_key: state.state_key().to_owned(),
                 content: state.content().into(),
             },
 
             Content::FailedToParseMessageLike { event_type, error } => {
-                TimelineItemContentKind::FailedToParseMessageLike {
+                TimelineItemContent::FailedToParseMessageLike {
                     event_type: event_type.to_string(),
                     error: error.to_string(),
                 }
             }
 
             Content::FailedToParseState { event_type, state_key, error } => {
-                TimelineItemContentKind::FailedToParseState {
+                TimelineItemContent::FailedToParseState {
                     event_type: event_type.to_string(),
                     state_key: state_key.to_string(),
                     error: error.to_string(),
                 }
             }
         }
     }
+}
 
-    pub fn as_message(self: Arc<Self>) -> Option<Arc<Message>> {
-        use matrix_sdk_ui::timeline::TimelineItemContent as Content;
-        unwrap_or_clone_arc_into_variant!(self, .0, Content::Message(msg) => Arc::new(Message(msg)))
+#[derive(Clone, uniffi::Record)]
+pub struct MessageContent {
+    pub msg_type: MessageType,
+    pub body: String,
+    pub in_reply_to: Option<Arc<InReplyToDetails>>,
+    pub thread_root: Option<String>,
+    pub is_edited: bool,
+    pub mentions: Option<Mentions>,
+}
+
+impl From<&matrix_sdk_ui::timeline::Message> for MessageContent {
```

Comment:
out of curiosity, could these be `From<Message> for MessageContent` instead? I'd suspect we could avoid lots of cloning in that case, and that might be sweet :)

Reply:
I don't think so unless I missed something important: either we clone a couple of values here or we need to clone the whole parent struct, or modify some other `From` and `TryFrom` implementations (like `impl TryFrom<MessageType> for RumaMessageType`) and clone other values there.

Reply:
I must be doing something wrong, because I end up with more cloning by removing the references in the `From` implementations, not less (see abe00c30709d294b1c0851f69d9e272f35a23a1f): most of these clones like `value.msgtype().clone().into()` happen because the functions have a signature like: `fn msgtype() -> &MessageType`, and I can't just use `(*value.msgtype()).into()` because I'd move the `value` variable.